### PR TITLE
Integrate Madrid Clinical Deterioration Workflow Into Pipeline

### PR DESCRIPTION
**Pull request by rathp**
Madrid dataset

# Prereqs

### Local Snakemake Installation
It's assumed you have a local install of Conda and Snakemake. It is also assumed that you have access to the Madrid Cluster.

### Pre-processed EHR Data
For instructions on pre-processing the raw data on the madrid cluster to get the demographics, vitals, labs and outcomes of the filtered cohort please refer to : [madrid-data-prep](https://github.com/tufts-ml/madrid-data-prep/tree/fix_preproc)

# Workflow

### SSH into Madrid VM
```
$ ssh to madrid vm
```

## A) Report Summary Statistics of Madrid EHR data
```
$ cd summary_statistics
$ snakemake --cores 1 --snakefile report_summary_statisitics.smk compute_summary_statistics
```

### 1) Filter admissions by patient-stay-slice (For eg, first 4 hrs, last 4hrs, first 30%)
---------------------------------------------------------------------------
Note : For eg. If we predict using first 4 hours of data, we ensure that every patient in the cohort has atleast 4 hours of data.
```
$ cd predictions_collapsed
$ snakemake --cores all --snakefile make_collapsed_dataset_per_tslice_and_split_train_test.smk filter_admissions_by_tslice_many_tslices
```
### 2) Collapsing features and saving to slice specific folders
----------------------------------------------------------------
```
$ snakemake --cores all --snakefile make_collapsed_dataset_per_tslice_and_split_train_test.smk collapse_features_many_tslices
```
### 3) Computing slice specific MEWS scores
--------------------------------------------
```
$ snakemake --cores all --snakefile make_collapsed_dataset_per_tslice_and_split_train_test.smk compute_mews_score_many_tslices
```

### 4) Merge all the collapsed features across tslices into a single features table
----------------------------------------------------------------------------------------
```
$ snakemake --cores 1 --snakefile make_collapsed_dataset_per_tslice_and_split_train_test.smk merge_collapsed_features_all_tslices
```

### 5) Split the features table into train - test. A single classifier will be trained on this training fold
-------------------------------------------------------------------------------------------------------------
```
$ snakemake --cores 1 --snakefile make_collapsed_dataset_per_tslice_and_split_train_test.smk split_into_train_and_test
```

### 6) Do every step above in squence
-------------------------------------------
```
$ snakemake --cores all --snakefile make_collapsed_dataset_per_tslice_and_split_train_test.smk all
```

### 7) Perform prediction of clinical deterioration with shallow models (LR, RF, MEWS) on multiple patient-stay-slices
```
$ cd predictions_collapsed
```

### 8) Predict with Logistic Regression
```
$ snakemake --cores 1 --snakefile logistic_regression.smk
```
### 9) Predict with Random Forest
```
$ snakemake --cores 1 --snakefile random_forest.smk
```

### 10) Predict with MEWS
```
$ snakemake --cores 1 --snakefile eval_mews_score.smk
```

### 11) Visualize prediction performance
```
$ cd predictions_collapsed
```
#### 11a) Plotting performance metrics such as AUROC, average precision, balanced accuracy on patient-stay-slice cohorts  
-------------------------------------------------------------------------------------------------------------
```
$ snakemake --cores 1 --snakefile evaluate_classifier_pertslice_performance.smk evaluate_performance
```

#### 11b) Plotting probability of deterioration over time for individual patient-stays
--------------------------------------------------------------------------------------------
```
$ snakemake --cores 1 --snakefile evaluate_proba_deterioration_over_time.smk evaluate_proba_deterioration
```

### 1) Filter and save multiple patient-stay-slices (0-8h, 0-16h, first 30%, last 5 hours) for evaluation
```
$ snakemake --cores all --snakefile make_features_and_outcomes_and_split_train_test.smk filter_admissions_by_tslice_many_tslices
```

### 2) Make a single set of features containing from all patient-stays and outcomes for each patient stay
```
$  snakemake --cores 1 --snakefile make_features_and_outcomes_and_split_train_test.smk make_features_and_outcomes
```
### 3) Split into train and test sets containing sequences
```
$ snakemake --cores 1 --snakefile make_features_and_outcomes_and_split_train_test.smk split_into_train_and_test
```
### 4) Impute missing values for train and test sets seaparately by carry forward and population mean
----------------------------------------------------------------------------------------------------------------
```
$ snakemake --cores 1 --snakefile make_features_and_outcomes_and_split_train_test.smk impute_missing_values
```
### 5) Predict with RNN with hyperparams specified in rnn.json
```
$ snakemake --cores all --snakefile rnn.smk train_and_evaluate_classifier_many_hyperparams
```

### CODE REVIEW

**Comment by michaelchughes** [comment]
Quick revisions requested @rathp

1) Would like to see the madrid/ folder have *same* file structure as other datasets.

See for example the EEG collapsed predictions scripts:
https://github.com/tufts-ml/time_series_prediction/tree/master/scripts/eeg/predictions_collapsed

2) For the collapsing:

* Can we avoid doing the "align to grid" imputation and just do straight collapsing?

I don't want to impute more than we have to. Imputation to per-hour levels to just then immediately collapse can really distort the statistics. 

* Why just slope and stddev? Why not at least mean, stddev, slope, missing, and last_measured?

This would take inspiration from the "simple imputation" strategy in the GRU-D paper.

missing = binary flag if value was missing in time period or not
hours_since_measured = number of hours in current sequence from the current timestep to when the value was last measured.   If no measurement available, should default to something large like 120 hours. 

Can you look up what "simple imputation" does if the value isn't available in the sequence? You could just fill with the population mean (which is zero IF you have already standardized that feature, but could be non-zero in general)

File: scripts/madrid/predictions_collapsed/Snakefile [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        os.path.join(MADRID_DATASET_STD_PATH, 'alignedfeatures_per_tstep.csv'),
+        os.path.join(MADRID_DATASET_STD_PATH, 'normfeatures_per_tstep.csv'),
+        os.path.join(MADRID_DATASET_STD_PATH, 'CollapsedFeaturesPerSequence.csv'),
+        os.path.join(MADRID_DATASET_STD_PATH, 'Spec_CollapsedFeaturesPerSequence.json'),	
+        os.path.join(MADRID_DATASET_STD_PATH, 'x_train.csv'),
+        os.path.join(MADRID_DATASET_STD_PATH, 'x_test.csv'),
+        os.path.join(MADRID_DATASET_STD_PATH, 'y_train.csv'),
+        os.path.join(MADRID_DATASET_STD_PATH, 'y_test.csv')
+
+rule align_to_grid:
```
Comment (michaelchughes) [inline scripts/madrid/predictions_collapsed/Snakefile:29]: @rathp Not sure we need to do "align to grid". Can do collapsing directly on original features

File: scripts/madrid/predictions_collapsed/Snakefile [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        '''
+        python -u {input.align_script} \
+            --input_ts_csv_path {input.x_csv} \
+            --data_dict {input.x_spec_json} \
+	    --step_size '1H' \
+            --output {output.aligned_csv} \
+        '''
+
+
+rule normalize_features:
```
Comment (michaelchughes) [inline scripts/madrid/predictions_collapsed/Snakefile:52]: OK to do this way for now.

In future, we want to use only the *training* set's features to do normalization. 
This uses all data (including test set)

File: scripts/madrid/predictions_collapsed/Snakefile [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+    shell:
+        '''
+        python -u {input.script} \
+            --input {input.x_csv} \
+            --data_dict {input.x_spec_json} \
+            --output {output.collapsedx_csv} \
+            --data_dict_output {output.collapsedx_json} \
+            --range_pairs "[(0,100)]" \
+            --collapse
```
Comment (michaelchughes) [inline scripts/madrid/predictions_collapsed/Snakefile:96]: This is where we want to make a change so we're creating a separate collapsed feature vector for each "prediction timestep" (e.g. every hour). That way we can properly evaluate a "risk score" rather than just a "given entire history, make one prediction" model

Probably want to be explicit about which summary features we are computing....

File: scripts/madrid/predictions_collapsed/logistic_regression/Snakefile [hunk 4]
```diff
@@ -0,0 +1,61 @@
+'''
```
Comment (michaelchughes) [inline scripts/madrid/predictions_collapsed/logistic_regression/Snakefile:1]: Please change this Snakefile to match filestructure of latest master branch.

See e.g. 
https://github.com/tufts-ml/time_series_prediction/tree/master/scripts/unimib_shar_activities/predictions_collapsed

Comment (michaelchughes) [reply scripts/madrid/predictions_collapsed/logistic_regression/Snakefile:1]: Most of the front env variable setting can be moved to a `config_loader.py` script

File: scripts/madrid/predictions_collapsed/Snakefile (hunk 1 above)
Comment (rathp) [reply scripts/madrid/predictions_collapsed/Snakefile]: Removed align_to_grid from feature collapsing workflow. Collapsing is now done using hours_from_admission as timestamps without any aligning to grid.

File: scripts/madrid/predictions_collapsed/Snakefile (hunk 2 above)
Comment (rathp) [reply scripts/madrid/predictions_collapsed/Snakefile]: Added normalization through sklearn Pipeline during training classifier to avoid leakage of information from validation/test into train.

File: scripts/madrid/predictions_collapsed/logistic_regression/Snakefile (hunk 4 above)
Comment (rathp) [reply scripts/madrid/predictions_collapsed/logistic_regression/Snakefile]: Changed file-structure to match : 
https://github.com/tufts-ml/time_series_prediction/tree/master/scripts/unimib_shar_activities/predictions_collapsed

Comment (rathp) [reply scripts/madrid/predictions_collapsed/logistic_regression/Snakefile]: Do we want all the data-prep scripts from madrid-data-prep in the time-series predict pipeline? I'm a bit confused about how to edit the config_loader.py. For now the front env variables are in the smk files.

**Comment by rathp** [comment]
> Quick revisions requested @rathp
> 
> 1. Would like to see the madrid/ folder have _same_ file structure as other datasets.
> 
> See for example the EEG collapsed predictions scripts:
> https://github.com/tufts-ml/time_series_prediction/tree/master/scripts/eeg/predictions_collapsed
> 
> 1. For the collapsing:
> 
> * Can we avoid doing the "align to grid" imputation and just do straight collapsing?
> 
> I don't want to impute more than we have to. Imputation to per-hour levels to just then immediately collapse can really distort the statistics.
- No more aligning to grid. Raw timestamps denoting hours_from_admission is used without any grid alignment for collapsing features 
> * Why just slope and stddev? Why not at least mean, stddev, slope, missing, and last_measured?
-  Added functions 'present' and 'hours_since_last measured' plus mean slope std median skew
> This would take inspiration from the "simple imputation" strategy in the GRU-D paper.
> 
> missing = binary flag if value was missing in time period or not
> hours_since_measured = number of hours in current sequence from the current timestep to when the value was last measured. If no measurement available, should default to something large like 120 hours.
- Done
> Can you look up what "simple imputation" does if the value isn't available in the sequence? You could just fill with the population mean (which is zero IF you have already standardized that feature, but could be non-zero in general)
- Yes imputation for a no-measurement case is replaced by population mean for that feature. In our case though, if no measurement is observed, feature is collapsed to 0.
